ssl/statem/extensions.c

```diff
@@ -0,0 +1,974 @@
+ * |context|. Returns 1 to indicate the extension is allowed or unknown or 0 to
+ * indicate the extension is not allowed. If returning 1 then |*found| is set to
+ * 1 if we found a definition for the extension, and |*idx| is set to its index
+ */
+static int verify_extension(SSL *s, unsigned int context, unsigned int type,
+                            custom_ext_methods *meths, int *found, size_t *idx)
+{
+    size_t i;
+    size_t builtin_num = OSSL_NELEM(ext_defs);
+
+    for (i = 0; i < builtin_num; i++) {
+        if (type == ext_defs[i].type) {
+            /* Check we're allowed to use this extension in this context */
+            if ((context & ext_defs[i].context) == 0)
+                return 0;
+
+            if (SSL_IS_DTLS(s)) {
+                if ((ext_defs[i].context & EXT_TLS_ONLY) != 0)
+                    return 0;
+            } else if ((ext_defs[i].context & EXT_DTLS_ONLY) != 0) {
+                    return 0;
+            }
+
+            *found = 1;
+            *idx = i;
+            return 1;
+        }
+    }
+
+    if ((context & (EXT_CLIENT_HELLO | EXT_TLS1_2_SERVER_HELLO)) == 0) {
+        /*
+         * Custom extensions only apply to <=TLS1.2. This extension is unknown
+         * in this context - we allow it
+         */
+        *found = 0;
+        return 1;
+    }
+
+    /* Check the custom extensions */
+    if (meths != NULL) {
+        for (i = builtin_num; i < builtin_num + meths->meths_count; i++) {
+            if (meths->meths[i - builtin_num].ext_type == type) {
+                *found = 1;
+                *idx = i;
+                return 1;
+            }
+        }
+    }
+
+    /* Unknown extension. We allow it */
+    *found = 0;
+    return 1;
+}
+
+/*
+ * Check whether the context defined for an extension |extctx| means whether
+ * the extension is relevant for the current context |thisctx| or not. Returns
+ * 1 if the extension is relevant for this context, and 0 otherwise
+ */
+static int extension_is_relevant(SSL *s, unsigned int extctx,
+                                 unsigned int thisctx)
+{
+    if ((SSL_IS_DTLS(s)
+                && (extctx & EXT_TLS_IMPLEMENTATION_ONLY) != 0)
+            || (s->version == SSL3_VERSION
+                    && (extctx & EXT_SSL3_ALLOWED) == 0)
+            || (SSL_IS_TLS13(s)
+                && (extctx & EXT_TLS1_2_AND_BELOW_ONLY) != 0)
+            || (!SSL_IS_TLS13(s) && (extctx & EXT_TLS1_3_ONLY) != 0))
+        return 0;
+
+    return 1;
+}
+
+/*
+ * Gather a list of all the extensions from the data in |packet]. |context|
+ * tells us which message this extension is for. The raw extension data is
+ * stored in |*res|. In the event of an error the alert type to use is stored in
+ * |*al|. We don't actually process the content of the extensions yet, except to
+ * check their types. This function also runs the initialiser functions for all
+ * known extensions (whether we have collected them or not).
```

Comment:
> collected? or found?

Changed to found.

> need to explain that |*res| should be OPENSSL_free'd if not null. and you should make sure to init it to null early on.

Well, no. |*res| only needs to be freed if the function returns successfully. Otherwise it is not set at all. I added a sentence to that effect.